app/components/autocomplete/slash_suggestion/apps_command_parser.ts

```diff
@@ -0,0 +1,802 @@
+
+    constructor(channelID = '', rootPostID = '') {
+        this.rootPostID = rootPostID;
+        this.channelID = channelID;
+        this.store = Store.redux as StoreType;
+    }
+
+    // decorateSuggestionComplete applies the necessary modifications for a suggestion to be processed
+    decorateSuggestionComplete = (pretext: string, choice: AutocompleteSuggestion): AutocompleteSuggestionWithComplete => {
+        // AutocompleteSuggestion.complete is required to be all text leading up to the current suggestion
+        const words = pretext.split(' ');
+        words.pop();
+        const before = words.join(' ') + ' ';
+
+        choice.hint = choice.hint || '';
+
+        let complete = before + (choice.complete || choice.suggestion);
+        complete = complete.substring(1);
+
+        const suggestion = {
+            ...choice,
+            suggestion: choice.suggestion,
+            complete,
+        };
+
+        return {
+            Complete: suggestion.complete,
+            Suggestion: suggestion.suggestion,
+            Hint: suggestion.hint || '',
+            Description: suggestion.description || '',
+            IconData: suggestion.iconData || '',
+        };
+    }
+
+    // getCommandBindings returns the commands in the redux store.
+    // They are grouped by app id since each app has one base command
+    getCommandBindings = (): AppBinding[] => {
+        const bindings = getAppsBindings(this.store.getState(), AppsBindings.COMMAND);
+        const grouped: {[appID: string]: AppBinding} = {};
+
+        for (const b of bindings) {
+            grouped[b.app_id] = grouped[b.app_id] || {
+                app_id: b.app_id,
+                label: b.app_id,
+                location: AppsBindings.COMMAND,
+                bindings: [],
+            };
+
+            const group = grouped[b.app_id];
+            group.bindings = group.bindings || [];
+            group.bindings.push(b);
+        }
+
+        return Object.values(grouped);
+    }
+
+    // getChannel computes the right channel, based on if this command is running in the center channel or RHS
+    getChannel = (): Channel | null => {
+        const state = this.store.getState();
+        if (!this.rootPostID && !this.channelID) {
+            return getCurrentChannel(state);
+        }
+
+        if (!this.rootPostID && this.channelID) {
+            return getChannel(state, this.channelID);
+        }
+
+        const post = getPost(state, this.rootPostID);
+        if (!post) {
+            return null;
+        }
+
+        return getChannel(state, post.channel_id);
+    }
+
+    // isAppCommand determines if subcommand/form suggestions need to be returned
+    isAppCommand = (pretext: string): boolean => {
+        for (const binding of this.getCommandBindings()) {
+            let base = binding.app_id;
+            if (!base) {
+                continue;
+            }
+
+            if (base[0] !== '/') {
```

Comment:
We currently don't have any restrictions on the app id. I suppose this check can be removed because if it did begin with a slash, we would want to prepend it with another slash anyway.